### PR TITLE
Updated English Courses and Playgrounds

### PR DESCRIPTION
**Pull request by pineapples5972**
## What does this PR do?
Add resource(s) | Remove resource(s) | Add info | Improve repo

## Checklist:
- [x] Read our [contributing guidelines](https://github.com/EbookFoundation/free-programming-books/blob/main/docs/CONTRIBUTING.md).
- [x] [Search](https://ebookfoundation.github.io/free-programming-books-search/) for duplicates.
- [x] Include author(s) and platform where appropriate.
- [x] Put lists in alphabetical order, correct spacing.
- [x] Add needed indications (PDF, access notes, under construction).
- [x] Used an informative name for this pull request.

## Follow-up

- Check the status of GitHub Actions and resolve any reported warnings!

### CODE REVIEW

**Comment by pineapples5972** [comment]
Finally All checks have passed.

File: courses/free-courses-en.md [hunk 1]
```diff
@@ -1245,7 +1250,9 @@
 * [Intro to Rust](https://www.youtube.com/playlist?list=PLJbE2Yu2zumDF6BX6_RdPisRVHgzV02NW) - Tensor Programming
 * [Learn Rust from scratch](https://www.educative.io/courses/learn-rust-from-scratch) - Educative.io
 * [Rust Basics](https://www.youtube.com/playlist?list=PLlcnQQJK8SUjApd95LIcd3K9XXmE-IeCS) - Engineer Man
+* [Rust Books](https://github.com/sger/RustBooks) - Books
```
Comment (shrn01) [inline courses/free-courses-en.md:1253]: Add direct links to each book

File: courses/free-courses-en.md [hunk 2]
```diff
@@ -1245,7 +1250,9 @@
...
 * [Rust Crash Course](https://www.youtube.com/watch?v=zF34dRivLOw) - Traversy Media
+* [Rust Learnings](https://github.com/ctjhoa/rust-learning) - a list of awesome rust learning material
```
Comment (shrn01) [inline courses/free-courses-en.md:1255]: Add each resource link directly

File: more/free-programming-playgrounds.md [hunk 3]
```diff
@@ -62,25 +63,33 @@
 * [Educational Visual CPU Simulator](https://github.com/Belotti01/CPU-Visual-Simulator) - Renato Cortinovis, Nicola Preda, Jonathan Cancelli, Alessandro Belotti, Davide Riva (JAVA, JAR)
 
 
+### <a id="bash"></a>Bash/Shell
+
+* [Bash Crawl](https://gitlab.com/slackermedia/bashcrawl)
+* [Terminal Quest](https://github.com/KanoComputing/terminal-quest)
```
Comment (shrn01) [inline more/free-programming-playgrounds.md:69]: Playgrounds are the ones which dont need any installation. Please remove any that require it

File: courses/free-courses-en.md [hunk 4]
```diff
@@ -1278,11 +1285,13 @@
 * [Introduction to Information Security](https://www.udacity.com/course/intro-to-information-security--ud459) - Wenke Lee, Mustaque Ahamad, Catherine Gamboa (Udacity)
 * [Learn Ethical Haacking From Scratch](https://www.udemy.com/course/learn-ethical-haacking-from-scratch) - Uzma Aslam (Udemy)
 * [Linux for Hackers and Pentesters](https://www.udemy.com/course/linux-for-hackers-and-pentesters) - Rajneesh Gupta (Udemy)
+* [Mobile Penetration Testing](https://selflearning.io/study-material/mobile-penetration-testing) - by selflearning.io
```
Comment (shrn01) [inline courses/free-courses-en.md:1288]: Looks like a book. Move all the books to the books file

**Review by shrn01 — CHANGES_REQUESTED** [review]

File: courses/free-courses-en.md (hunk 2 above)
Comment (pineapples5972) [reply courses/free-courses-en.md:1255]: Rust learning is seperate repo for listing out books, videos and various resources on learning rust.
Its impossible to update this repo with that all the time. and as this repo doesnt have a seperate list for repos or other resources for this to fit in i'll just remove this line from here.

File: more/free-programming-playgrounds.md (hunk 3 above)
Comment (pineapples5972) [reply more/free-programming-playgrounds.md:69]: to bad we dont have programs or games category for this.

File: books/free-programming-books-langs.md [hunk 5]
```diff
@@ -2248,11 +2248,12 @@ That section got so big, we decided to split it into its own file, the [BY SUBJE
 * [Easy Rust](https://dhghomon.github.io/easy_rust/) - David McLeod (HTML, PDF)
 * [From JavaScript to Rust ebook](https://github.com/wasmflow/node-to-rust/raw/HEAD/from-javascript-to-rust.pdf) - Jarrod Overson (PDF)
 * [Guide to Rustc Development](https://rustc-dev-guide.rust-lang.org)
-* [Learn Rust in Y minutes](https://learnxinyminutes.com/docs/rust/)
+* [Learn Rust in Y minutes](https://learnxinyminutes.com/docs/rust)
```
Comment (eshellman) [inline books/free-programming-books-langs.md:2251]:
```suggestion
* [Learn Rust in Y minutes](https://learnxinyminutes.com/docs/rust/)
```
trailing slash only needed for website root

File: books/free-programming-books-langs.md [hunk 6]
```diff
@@ -2263,46 +2264,47 @@ That section got so big, we decided to split it into its own file, the [BY SUBJE
...
 * [The Rustonomicon](https://doc.rust-lang.org/nomicon)
+* [Welcome to Rust-101](https://www.ralfj.de/projects/rust-101/main.html)
 * [Why Rust?](https://www.oreilly.com/content/why-rust)
 
 
 ### Sage
 
 * [Sage for Power Users](http://wstein.org/books/sagebook/sagebook.pdf) - William Stein (PDF)
-* [The Sage Manuals](http://www.sagemath.org/doc/)
+* [The Sage Manuals](http://www.sagemath.org/doc)
```
Comment (eshellman) [inline books/free-programming-books-langs.md:2274]:
```suggestion
* [The Sage Manuals](http://www.sagemath.org/doc/)
```

File: more/free-programming-playgrounds.md [hunk 7]
```diff
@@ -110,7 +111,7 @@
 
 ### DevOps
 
-* [KodeKloud](https://kodekloud.com/kodekloud-engineer/)
+* [KodeKloud](https://kodekloud.com/kodekloud-engineer)
```
Comment (eshellman) [inline more/free-programming-playgrounds.md:114]:
```suggestion
* [KodeKloud](https://kodekloud.com/kodekloud-engineer/)
```

File: more/free-programming-playgrounds.md [hunk 8]
```diff
@@ -64,23 +64,24 @@
...
+* [JDoodle - Online C Compiler IDE](https://www.jdoodle.com/c-online-compiler)
+* [Programiz - Online C Compiler](https://www.programiz.com/c-programming/online-compiler)
 * [SoloLearn](https://code.sololearn.com/c)
 
 
 ### <a id="csharp"></a>C\#
 
-* [Programiz - Online Compiler](https://www.programiz.com/csharp-programming/online-compiler/)
+* [codebattle](https://codebattle.hexlet.io)
+* [Programiz - Online Compiler](https://www.programiz.com/csharp-programming/online-compiler)
```
Comment (eshellman) [inline more/free-programming-playgrounds.md:76]:
```suggestion
* [Programiz - Online Compiler](https://www.programiz.com/csharp-programming/online-compiler/)
```

File: more/free-programming-playgrounds.md [hunk 9]
```diff
@@ -64,23 +64,24 @@
 
 ### <a id="c"></a>C
 
-* [InterviewBit - Online C Compiler IDE](https://www.interviewbit.com/online-c-compiler/)
-* [JDoodle - Online C Compiler IDE](https://www.jdoodle.com/c-online-compiler/)
-* [Programiz - Online C Compiler](https://www.programiz.com/c-programming/online-compiler/)
+* [InterviewBit - Online C Compiler IDE](https://www.interviewbit.com/online-c-compiler)
```
Comment (eshellman) [inline more/free-programming-playgrounds.md:67]:
```suggestion
* [InterviewBit - Online C Compiler IDE](https://www.interviewbit.com/online-c-compiler/)
```

File: more/free-programming-playgrounds.md [hunk 10]
```diff
@@ -64,23 +64,24 @@
...
+* [JDoodle - Online C Compiler IDE](https://www.jdoodle.com/c-online-compiler)
```
Comment (eshellman) [inline more/free-programming-playgrounds.md:68]:
```suggestion
* [JDoodle - Online C Compiler IDE](https://www.jdoodle.com/c-online-compiler/)
```

File: books/free-programming-books-langs.md [hunk 11]
```diff
@@ -2263,46 +2264,47 @@ That section got so big, we decided to split it into its own file, the [BY SUBJE
...
+* [PythonToScala](https://wrobstory.gitbooks.io/python-to-scala/content) - Rob Story
+* [S-99: Ninety-Nine Scala Problems](http://aperiodic.net/phil/scala/s-99) - Phil! Gold
 * [Scala & Design Patterns: Exploring Language Expressivity](http://www.scala-lang.org/old/sites/default/files/FrederikThesis.pdf) - Fredrik Skeel Løkke (PDF)
-* [Scala Book](https://alvinalexander.com/scala/scala-book-free/) - Alvin Alexander (PDF, MOBI, HTML, EPUB)
+* [Scala Book](https://alvinalexander.com/scala/scala-book-free) - Alvin Alexander (PDF, MOBI, HTML, EPUB)
 * [Scala By Example](https://www.scala-lang.org/old/sites/default/files/linuxsoft_archives/docu/files/ScalaByExample.pdf) - M. Odersky (PDF)
 * [Scala Cookbook: Bonus Chapters](http://examples.oreilly.com/9781449339616-files/Scala_Cookbook_bonus_chapters.pdf) - Alvin Alexander (PDF)
 * [Scala for Perl 5 Programmers](https://github.com/garu/scala-for-perl5-programmers) - Breno G. de Oliveira
-* [Scala School by Twitter](http://twitter.github.io/scala_school/)
+* [Scala School by Twitter](http://twitter.github.io/scala_school)
```
Comment (eshellman) [inline books/free-programming-books-langs.md:2300]:
```suggestion
* [Scala School by Twitter](http://twitter.github.io/scala_school/)
```

File: books/free-programming-books-langs.md [hunk 12]
```diff
@@ -2263,46 +2264,47 @@ That section got so big, we decided to split it into its own file, the [BY SUBJE
...
+* [Pro Scala: Monadic Design Patterns for the Web](https://github.com/leithaus/XTrace/tree/monadic/src/main/book/content)
+* [Programming in Scala, First Edition](http://www.artima.com/pins1ed) - M. Odersky, L. Spoon, B. Venners
 * [Pure functional HTTP APIs in Scala](https://leanpub.com/pfhais/read) - Jens Grassel
-* [PythonToScala](https://wrobstory.gitbooks.io/python-to-scala/content/) - Rob Story
-* [S-99: Ninety-Nine Scala Problems](http://aperiodic.net/phil/scala/s-99/) - Phil! Gold
+* [PythonToScala](https://wrobstory.gitbooks.io/python-to-scala/content) - Rob Story
+* [S-99: Ninety-Nine Scala Problems](http://aperiodic.net/phil/scala/s-99) - Phil! Gold
 * [Scala & Design Patterns: Exploring Language Expressivity](http://www.scala-lang.org/old/sites/default/files/FrederikThesis.pdf) - Fredrik Skeel Løkke (PDF)
-* [Scala Book](https://alvinalexander.com/scala/scala-book-free/) - Alvin Alexander (PDF, MOBI, HTML, EPUB)
+* [Scala Book](https://alvinalexander.com/scala/scala-book-free) - Alvin Alexander (PDF, MOBI, HTML, EPUB)
```
Comment (eshellman) [inline books/free-programming-books-langs.md:2296]:
```suggestion
* [Scala Book](https://alvinalexander.com/scala/scala-book-free/) - Alvin Alexander (PDF, MOBI, HTML, EPUB)
```

File: books/free-programming-books-langs.md [hunk 13]
```diff
@@ -2263,46 +2264,47 @@ That section got so big, we decided to split it into its own file, the [BY SUBJE
...
-* [Programming in Scala, First Edition](http://www.artima.com/pins1ed/) - M. Odersky, L. Spoon, B. Venners
+* [Learning Scala in small bites](http://matt.might.net/articles/learning-scala-in-small-bites)
+* [Learning Scalaz](http://eed3si9n.com/learning-scalaz)
+* [Pro Scala: Monadic Design Patterns for the Web](https://github.com/leithaus/XTrace/tree/monadic/src/main/book/content)
+* [Programming in Scala, First Edition](http://www.artima.com/pins1ed) - M. Odersky, L. Spoon, B. Venners
 * [Pure functional HTTP APIs in Scala](https://leanpub.com/pfhais/read) - Jens Grassel
-* [PythonToScala](https://wrobstory.gitbooks.io/python-to-scala/content/) - Rob Story
-* [S-99: Ninety-Nine Scala Problems](http://aperiodic.net/phil/scala/s-99/) - Phil! Gold
+* [PythonToScala](https://wrobstory.gitbooks.io/python-to-scala/content) - Rob Story
+* [S-99: Ninety-Nine Scala Problems](http://aperiodic.net/phil/scala/s-99) - Phil! Gold
```
Comment (eshellman) [inline books/free-programming-books-langs.md:2294]:
```suggestion
* [S-99: Ninety-Nine Scala Problems](http://aperiodic.net/phil/scala/s-99/) - Phil! Gold
```

File: books/free-programming-books-langs.md (hunk 5 above)
Comment (pineapples5972) [reply books/free-programming-books-langs.md:2251]: ooh sorry i didnt knew

**Comment by eshellman** [comment]
I suggest starting a new PR from a fresh branch. Or 3 PRs, one for playgrounds, one for books, and on for courses. This will make it easier for us to review and merge your changes. Thanks so much for persisting - our hacktoberfest cleanup is mostly to blame for the conflicts.

**Comment by pineapples5972** [comment]
> I suggest starting a new PR from a fresh branch. 

I am afraid I am too overwhelmed by this point to do that.
I even lost track of what links I was including in the first place.

> Or 3 PRs, one for playgrounds, one for books, and on for courses. This will make it easier for us to review and merge your changes.

I'll consider this for next PRs if I ever get enough energy and time to contribute maybe in weekends or so.
Thanks tho for keeping up with my naive mistakes.

**Pull request closed without merging**